pkg/integrations/v2/eventhandler/eventhandler.go

```diff
@@ -0,0 +1,421 @@
+	if !ok {
+		level.Error(eh.Log).Log("msg", "Object not v1.Event", "obj", obj)
+		return
+	}
+	err := eh.handleEvent(event)
+	if err != nil {
+		level.Error(eh.Log).Log("msg", "Error handling event", "err", err, "event", event)
+	}
+}
+
+// Handles event object updates. Note that this get triggered on informer resyncs and also
+// events occurring more than once (in which case .count is incremented)
+func (eh *EventHandler) updateEvent(objOld interface{}, objNew interface{}) {
+	eOld, ok := objOld.(*v1.Event)
+	if !ok {
+		level.Error(eh.Log).Log("msg", "Object not v1.Event", "obj", eOld)
+		return
+	}
+
+	eNew, ok := objNew.(*v1.Event)
+	if !ok {
+		level.Error(eh.Log).Log("msg", "Object not v1.Event", "obj", eNew)
+		return
+	}
+
+	if eOld.GetResourceVersion() == eNew.GetResourceVersion() {
+		// ignore resync updates
+		level.Debug(eh.Log).Log("msg", "Event RV didn't change, ignoring", "eRV", eNew.ResourceVersion)
+		return
+	}
+	err := eh.handleEvent(eNew)
+	if err != nil {
+		level.Error(eh.Log).Log("msg", "Error handling event", "err", err, "event", eNew)
+	}
+}
+
+func (eh *EventHandler) handleEvent(event *v1.Event) error {
+	eventTs := getTimestamp(event)
+
+	// if event is older than the one stored in cache on startup, we've shipped it
+	if eventTs.Before(eh.InitEvent.Timestamp) {
+		return nil
+	}
+	// if event is equal and is in map, we've shipped it
+	if eventTs.Equal(eh.InitEvent.Timestamp) {
+		if _, ok := eh.InitEvent.RvMap[event.ResourceVersion]; ok {
+			return nil
+		}
+	}
+
+	labels, msg, err := eh.extractEvent(event)
+	if err != nil {
+		return err
+	}
+
+	entry := newEntry(msg, eventTs, labels)
+	ok := eh.LogsClient.Instance(eh.LogsInstance).SendEntry(entry, eh.SendTimeout)
+	if !ok {
+		err = fmt.Errorf("msg=%s entry=%s", "error handing entry off to promtail", entry)
+		return err
+	}
+	level.Info(eh.Log).Log("msg", "Shipped entry", "eventRV", event.ResourceVersion, "eventMsg", event.Message)
+
+	// update cache with new "last" event
+	err = eh.updateLastEvent(event, eventTs)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (eh *EventHandler) deleteEvent(obj interface{}) {
+	// do nothing...
+	// todo: log this maybe??
+}
+
+// extract data from event fields and create labels, etc.
+// TODO: ship JSON blobs and allow users to configure using pipelines etc.
+// instead of hardcoding labels here
+func (eh *EventHandler) extractEvent(event *v1.Event) (model.LabelSet, string, error) {
+	var (
+		msg    strings.Builder
+		labels = make(model.LabelSet)
+	)
+
+	obj := event.InvolvedObject
+	if obj.Name == "" {
+		return nil, "", fmt.Errorf("no involved object for event")
+	}
+
+	labels[model.LabelName("source")] = model.LabelValue("eventhandler")
+	labels[model.LabelName("cluster")] = model.LabelValue(eh.ClusterLabel)
+
+	kindStr := strings.ToLower(obj.Kind)
+	// TODO: match up labels with k8s integration so that log / metric correlation works
+	// and users can use the same dashboard for both logs and metrics
```

Comment:
Wouldn't this TODO be handled when you implement the ability to ship JSON blobs and allow pipelines?

That, and/or allowing label rewrites for the integration would allow the user to decide how they want to take the "source" labels and turn them into the log labels.

In fact, couldn't/shouldn't these be `__meta__` labels like metrics or logs fetched by a `kubernetes_sd_config`, allowing the user to manipulate them as desired before coming to rest in cortex/loki?

Reply:
Yep exactly, that TODO would be eliminated with pipeline/relabel etc. support - it's more just a temp reminder to match things up in case we want to get correlation working for the k8s integration